File: frontend/src/components/extractor.js
/**
 * extractor.js
 * ============
 * Rule-based NLP extraction. Ported from backend/extractor.py so the app
 * runs entirely in the browser with no server dependency.
 *
 * Takes a transcript string and returns:
 *   { complaints, diagnosis, medicines[], labs[], notes }
 */

// --------------------------------------------------------------------------
// Reference data
// --------------------------------------------------------------------------
const KNOWN_LABS = [
  'CBC', 'ECG', 'X-Ray', 'Chest X-Ray', 'RBS', 'FBS', 'PPBS', 'HbA1c',
  'Lipid Profile', 'LFT', 'KFT', 'RFT', 'MP', 'Widal', 'COVID-19 RTPCR',
  'Urine Routine', 'Urine RE', 'TSH', 'T3', 'T4', 'Vitamin D', 'Vitamin B12',
  'USG', 'USG Abdomen', 'CT Scan', 'MRI', 'Echo', 'TMT', 'Holter',
  'Blood Sugar', 'ESR', 'CRP', 'Dengue NS1', 'Typhidot',
]

const FREQUENCY_MAP = {
  'od': '1-0-0', 'once daily': '1-0-0', 'once a day': '1-0-0',
  'bd': '1-0-1', 'bid': '1-0-1', 'twice daily': '1-0-1', 'twice a day': '1-0-1',
  'tid': '1-1-1', 'tds': '1-1-1', 'three times a day': '1-1-1',
  'qid': '1-1-1-1', 'four times a day': '1-1-1-1',
  'hs': '0-0-1', 'at bedtime': '0-0-1', 'at night': '0-0-1',
  'sos': 'SOS', 'prn': 'SOS', 'as needed': 'SOS',
}

const COMMON_MEDS = [
  'paracetamol', 'ibuprofen', 'amoxicillin', 'azithromycin', 'amlodipine',
  'atorvastatin', 'metformin', 'pantoprazole', 'omeprazole', 'cetirizine',
  'montelukast', 'telmisartan', 'losartan', 'ramipril', 'metoprolol',
  'ciprofloxacin', 'levofloxacin', 'cefixime', 'ondansetron', 'domperidone',
  'diclofenac', 'aceclofenac', 'tramadol', 'clopidogrel', 'aspirin',
  'salbutamol', 'prednisolone', 'dexamethasone', 'insulin', 'levothyroxine',
  'crocin', 'dolo', 'calpol', 'combiflam', 'azee', 'augmentin', 'zifi',
  'pan', 'pantocid', 'nexpro', 'rabeprazole', 'telma', 'stamlo',
]

const MED_SUFFIXES = [
  'cillin', 'mycin', 'azole', 'pril', 'sartan', 'olol',
  'statin', 'dipine', 'prazole', 'floxacin',
]

// --------------------------------------------------------------------------
// Helpers
// --------------------------------------------------------------------------
function escapeRegex(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function looksLikeMedicine(word) {
  const w = word.toLowerCase()
  return MED_SUFFIXES.some((s) => w.endsWith(s))
}

function findFrequencyAlias(text, medName) {
  const idx = text.indexOf(medName)
  if (idx === -1) return ''
  const window = text.slice(idx, idx + 80)
  for (const [alias, code] of Object.entries(FREQUENCY_MAP)) {
    const re = new RegExp('\\b' + escapeRegex(alias) + '\\b', 'i')
    if (re.test(window)) return code
  }
  return ''
}

// --------------------------------------------------------------------------
// Field extractors
// --------------------------------------------------------------------------
function extractComplaints(text) {
  const patterns = [
    /(?:presenting with|presents with|complaints? of|c\/?o|chief complaints?:?-?)\s+([^.;]+?)(?:\.|;|diagnosed|diagnosis|assessment|impression|prescrib|advis|$)/i,
    /patient (?:has|is having|reports?)\s+([^.;]+?)(?:\.|;|since|for|$)/i,
  ]
  for (const pat of patterns) {
    const m = text.match(pat)
    if (m) return m[1].trim().replace(/[,.]$/, '')
  }
  return ''
}

function extractDiagnosis(text) {
  const patterns = [
    /(?:diagnos(?:ed|is)(?:\s+(?:with|of|is))?|assessment:?-?|impression:?-?)\s+([^.;]+?)(?:\.|;|prescrib|advis|medicines?|tests?|labs?|$)/i,
    /likely\s+([^.;]+?)(?:\.|;|$)/i,
  ]
  for (const pat of patterns) {
    const m = text.match(pat)
    if (m) return m[1].trim().replace(/[,.]$/, '')
  }
  return ''
}

function extractMedicines(text) {
  const medicines = []
  const textLower = text.toLowerCase()
  const seen = new Set()

  // Pattern: name + dosage + optional frequency-code + optional duration
  const re = /\b([a-zA-Z]{4,})\s+(\d+\s*(?:mg|ml|mcg|g))\s*(\d-\d-\d(?:-\d)?)?\s*(?:for\s+(\d+\s+(?:day|week|month)s?))?/gi
  let m
  while ((m = re.exec(text)) !== null) {
    const name = m[1]
    const nameLower = name.toLowerCase()
    if (!COMMON_MEDS.includes(nameLower) && !looksLikeMedicine(nameLower)) continue
    if (seen.has(nameLower)) continue
    seen.add(nameLower)

    const dosage = m[2] || ''
    const freq = m[3] || findFrequencyAlias(textLower, nameLower)
    const duration = m[4] || ''
    medicines.push({
      name: name.charAt(0).toUpperCase() + name.slice(1).toLowerCase() + (dosage ? ' ' + dosage : ''),
      dosage,
      frequency: freq,
      duration,
    })
  }

  // Fallback: scan COMMON_MEDS by name even without explicit dosage
  if (medicines.length === 0) {
    for (const med of COMMON_MEDS) {
      if (textLower.includes(med) && !seen.has(med)) {
        const dosageRe = new RegExp(escapeRegex(med) + '\\s+(\\d+\\s*(?:mg|ml|mcg|g))', 'i')
        const durRe = new RegExp(escapeRegex(med) + '.*?for\\s+(\\d+\\s+(?:day|week|month)s?)', 'i')
        const dM = textLower.match(dosageRe)
        const durM = textLower.match(durRe)
        const display = med.charAt(0).toUpperCase() + med.slice(1) + (dM ? ' ' + dM[1] : '')
        medicines.push({
          name: display,
          dosage: dM ? dM[1] : '',
          frequency: findFrequencyAlias(textLower, med),
          duration: durM ? durM[1] : '',
        })
        seen.add(med)
      }
    }
  }

  return medicines
}

function extractLabs(text) {
  const found = []
  const textLower = text.toLowerCase()
  for (const lab of KNOWN_LABS) {
    if (textLower.includes(lab.toLowerCase()) && !found.includes(lab)) {
      found.push(lab)
    }
  }
  return found
}

function extractNotes(text) {
  const pat = /(?:advis(?:e|ed)|notes?|instruction[s]?|follow[- ]?up):?-?\s+([^.;]+?)(?:\.|;|$)/i
  const m = text.match(pat)
  return m ? m[1].trim().replace(/[,.]$/, '') : ''
}

// --------------------------------------------------------------------------
// Main entry
// --------------------------------------------------------------------------
export function extract(transcript) {
  if (!transcript) {
    return { complaints: '', diagnosis: '', medicines: [], labs: [], notes: '' }
  }
  return {
    complaints: extractComplaints(transcript),
    diagnosis: extractDiagnosis(transcript),
    medicines: extractMedicines(transcript),
    labs: extractLabs(transcript),
    notes: extractNotes(transcript),
  }
}

File: backend/extractor.py
"""
extractor.py
============
Rule-based NLP extraction engine.

Takes raw transcript text and extracts structured medical fields using
regex and keyword matching. Runs entirely offline — no external APIs.
"""

import re

# --------------------------------------------------------------------------
# Reference data
# --------------------------------------------------------------------------
KNOWN_LABS = [
    "CBC", "ECG", "X-Ray", "Chest X-Ray", "RBS", "FBS", "PPBS", "HbA1c",
    "Lipid Profile", "LFT", "KFT", "RFT", "MP", "Widal", "COVID-19 RTPCR",
    "Urine Routine", "Urine RE", "TSH", "T3", "T4", "Vitamin D", "Vitamin B12",
    "USG", "USG Abdomen", "CT Scan", "MRI", "Echo", "TMT", "Holter",
    "Blood Sugar", "ESR", "CRP", "Dengue NS1", "Typhidot",
]

FREQUENCY_MAP = {
    "od": "1-0-0", "once daily": "1-0-0", "once a day": "1-0-0",
    "bd": "1-0-1", "bid": "1-0-1", "twice daily": "1-0-1", "twice a day": "1-0-1",
    "tid": "1-1-1", "tds": "1-1-1", "three times a day": "1-1-1",
    "qid": "1-1-1-1", "four times a day": "1-1-1-1",
    "hs": "0-0-1", "at bedtime": "0-0-1", "at night": "0-0-1",
    "sos": "SOS", "prn": "SOS", "as needed": "SOS",
}

COMMON_MEDS = [
    "paracetamol", "ibuprofen", "amoxicillin", "azithromycin", "amlodipine",
    "atorvastatin", "metformin", "pantoprazole", "omeprazole", "cetirizine",
    "montelukast", "telmisartan", "losartan", "ramipril", "metoprolol",
    "ciprofloxacin", "levofloxacin", "cefixime", "ondansetron", "domperidone",
    "diclofenac", "aceclofenac", "tramadol", "clopidogrel", "aspirin",
    "salbutamol", "prednisolone", "dexamethasone", "insulin", "levothyroxine",
    "crocin", "dolo", "calpol", "combiflam", "azee", "augmentin", "zifi",
    "pan", "pantocid", "nexpro", "rabeprazole", "telma", "stamlo",
]


# --------------------------------------------------------------------------
# Extraction functions
# --------------------------------------------------------------------------
def extract_complaints(text):
    """Find phrases after 'presenting with', 'complaints of', 'c/o'."""
    patterns = [
        r"(?:presenting with|presents with|complaints? of|c/?o|chief complaints?[:\-]?)\s+([^.;]+?)(?:\.|;|diagnosed|diagnosis|assessment|impression|prescrib|advis|$)",
        r"patient (?:has|is having|reports?)\s+([^.;]+?)(?:\.|;|since|for|$)",
    ]
    for pat in patterns:
        m = re.search(pat, text, re.IGNORECASE)
        if m:
            return m.group(1).strip(" ,.")
    return ""


def extract_diagnosis(text):
    """Find phrases after 'diagnosed with', 'assessment', 'impression'."""
    patterns = [
        r"(?:diagnos(?:ed|is)(?:\s+(?:with|of|is))?|assessment[:\-]?|impression[:\-]?)\s+([^.;]+?)(?:\.|;|prescrib|advis|medicines?|tests?|labs?|$)",
        r"likely\s+([^.;]+?)(?:\.|;|$)",
    ]
    for pat in patterns:
        m = re.search(pat, text, re.IGNORECASE)
        if m:
            return m.group(1).strip(" ,.")
    return ""


def extract_medicines(text):
    """
    Extract medicine entries.
    Looks for: name + optional dosage + optional frequency + optional duration.
    """
    medicines = []
    text_lower = text.lower()

    # Pattern: medicine name with dosage and timing
    # e.g. "Paracetamol 500mg 1-0-1 for 5 days"
    pattern_full = re.compile(
        r"\b([a-z]{4,})\s+(\d+\s*(?:mg|ml|mcg|g))\s*(?:(\d-\d-\d(?:-\d)?))?\s*(?:for\s+(\d+\s+(?:day|week|month)s?))?",
        re.IGNORECASE,
    )

    for m in pattern_full.finditer(text):
        name = m.group(1)
        dosage = m.group(2) or ""
        freq = m.group(3) or ""
        duration = m.group(4) or ""

        # Filter out non-medicine words
        if name.lower() not in COMMON_MEDS and not _looks_like_medicine(name):
            continue

        medicines.append({
            "name": name.title() + (f" {dosage}" if dosage else ""),
            "dosage": dosage,
            "frequency": freq or _find_frequency_alias(text_lower, name.lower()),
            "duration": duration,
        })

    # Pattern: medicine name with frequency alias (BD, TID, etc.) and no explicit dosage
    if not medicines:
        for med in COMMON_MEDS:
            if med in text_lower:
                dosage_match = re.search(
                    re.escape(med) + r"\s+(\d+\s*(?:mg|ml|mcg|g))",
                    text_lower,
                )
                dur_match = re.search(
                    re.escape(med) + r".*?for\s+(\d+\s+(?:day|week|month)s?)",
                    text_lower,
                )
                medicines.append({
                    "name": med.title() + (f" {dosage_match.group(1)}"
                                           if dosage_match else ""),
                    "dosage": dosage_match.group(1) if dosage_match else "",
                    "frequency": _find_frequency_alias(text_lower, med),
                    "duration": dur_match.group(1) if dur_match else "",
                })

    return medicines


def extract_labs(text):
    """Match known lab test names."""
    found = []
    text_lower = text.lower()
    for lab in KNOWN_LABS:
        if lab.lower() in text_lower and lab not in found:
            found.append(lab)
    return found


def extract_notes(text):
    """Find advice / notes section."""
    patterns = [
        r"(?:advis(?:e|ed)|notes?|instruction[s]?|follow[- ]?up)[:\-]?\s+([^.;]+?)(?:\.|;|$)",
    ]
    for pat in patterns:
        m = re.search(pat, text, re.IGNORECASE)
        if m:
            return m.group(1).strip(" ,.")
    return ""


# --------------------------------------------------------------------------
# Helpers
# --------------------------------------------------------------------------
def _looks_like_medicine(word):
    """Heuristic: medicines often end in common suffixes."""
    suffixes = ("cillin", "mycin", "azole", "pril", "sartan", "olol",
                "statin", "dipine", "prazole", "floxacin")
    return any(word.lower().endswith(s) for s in suffixes)


def _find_frequency_alias(text, med_name):
    """Look for frequency phrase within ~50 chars after the medicine name."""
    idx = text.find(med_name)
    if idx == -1:
        return ""
    window = text[idx:idx + 80]
    for alias, code in FREQUENCY_MAP.items():
        if re.search(r"\b" + re.escape(alias) + r"\b", window):
            return code
    return ""


# --------------------------------------------------------------------------
# Main entry point
# --------------------------------------------------------------------------
def extract(transcript):
    """Run the full extraction pipeline on a transcript."""
    if not transcript:
        return {
            "complaints": "",
            "diagnosis": "",
            "medicines": [],
            "labs": [],
            "notes": "",
        }

    return {
        "complaints": extract_complaints(transcript),
        "diagnosis": extract_diagnosis(transcript),
        "medicines": extract_medicines(transcript),
        "labs": extract_labs(transcript),
        "notes": extract_notes(transcript),
    }
